### DISCUSSION TITLE
How do I actually use incremental parsing, querying, etc?

### DISCUSSION THREAD

**Discussion by mauricioszabo** [tree-sitter/tree-sitter#2057] [Q&A] [answered]
I am doing some experimentation on web-tree-sitter for the Pulsar editor. It's going ok, but there are some cases I need to work on. The current problem I'm facing is on knowing what exactly changed. So for example, I'm using the Ruby parser:

```js
const tree = parser.parse("def foobar\n  1\nend")
// I want to make the "1" become "10"
tree.edit({
  startPosition: {row: 1, column: 3},
  oldEndPosition: {row: 1, column: 3},
  newEndPosition: {row: 1, column: 4},
  startIndex: 14,
  oldEndIndex: 14,
  newEndIndex: 15
})

const newTree = parser.parse("def foobar\n  10\nend", tree)
tree.getChangedRanges(newTree)
```

The `getChangedRanges` returns `[]`, which basically is not correct - on `tree` (before running `.edit`) the function `tree.rootNode.children[0].children[2]` returns a node that starts on `{row: 1, column: 2}` and ends on `{row: 1, column: 3}`; but `newTree.rootNode.children[0].children[2]` ends on `{row: 1, column: 4}`

The problem with this is that I can't reliably query only what was modified, to basically invalidate the old highlighting and ask the editor to re-validate that part.

So, how should I do this? I noticed that just the fact that I called `.edit` _actually changed_ the `endPosition` to `{row: 1, column: 4}` and made the `.text` to be `"1\n"`. Is this a bug?

**Comment by maxbrunsfeld** [comment, marked as answer]
Yeah, the purpose of `edit` is to reshape a syntax tree to reflect an edit to the code.

The `getChangedRanges` function returns the ranges where the syntax has changed between the (edited) old tree and the new tree. So in your case, because you changed `1` to `10`, which is still an integer literal, there are no areas that have changed syntax.

**Reply by maxbrunsfeld** [reply]
I do think there's a bug where `.text` returns the wrong thing after a tree is edited. There's actually no way for it to return the *right* thing, since you haven't told Tree-sitter the new text yet. We should probably just return `null` if the node is edited, to remind you that the new text isn't known to tree-sitter yet.

**Reply by mauricioszabo** [reply]
Sure, but how do I detect that my syntax now _ends_ in column `4` and not on `3`? Is there anything I can do? Otherwise, the editor highlights `1` as number, and `0` gets no info on what it actually is...

**Reply by maxbrunsfeld** [reply]
Your editor needs to re-render lines that have changed. Tree-sitter doesn't record that text-related state for you. It just provides the core ability to detect which ranges have changed their *syntactic* structure after a change.

**Reply by mauricioszabo** [reply]
Or should I just send to `.matches` the actual range that I changed, together with what `getChangedRanges` return, and update the editor with everything that's returned for each query?

**Reply by danielpunkass** [reply]
Isn't it true that `ts_tree_get_changed_ranges` (which I assume the `getChangedRanges` binding here calls through to) does more than to report changes in the pure syntax of a tree? The documentation suggests calling it not after the tree edit, but after the subsequent parse. I think that, for example, any subtree having changed size would constitute a change as far as it's concerned, so that subtree's range would be added to the list of changed ranges.

In the case described here wouldn't the code around here: https://github.com/tree-sitter/tree-sitter/blob/584b55df8d1439833c58003663a5e0ec862905e0/lib/src/get_changed_ranges.c#L491 end up determining that subtree containing `1`, which changed to `10`, had doubled in size an thus assign its range to the list of changed ranges?

**Reply by danielpunkass** [reply]
I think that possibly in the original code sample you want to call:

```
newTree.getChangedRanges(tree)
```

Does that make a change in whether it reports changed ranges? I'm pretty sure it's more conventional for the old tree to be considered the "reference" against which changes in the new tree are calculated. Worth a shot anyway!

**Reply by mauricioszabo** [reply]
@danielpunkass I also did try that, same result. Also, I was following the tests of tree-sitter-web - on these tests, getChangedRanges is, indeed, called on the old tree and the new tree is passed as parameter...

**Reply by danielpunkass** [reply]
@mauricioszabo Ah well! I do think though that your expectations about how `getChangedRanges` should work is correct. I'll be curious to hear if @maxbrunsfeld has another perspective in light of the code I pointed to.
